Assigment2/controller/controller.rb

```diff
@@ -0,0 +1,143 @@
+require_relative '../repository/transaction_accounts_repository'
+require_relative '../repository/accounts_repository'
+require_relative '../repository/ATMs_repository'
+require_relative '../repository/transaction_ATM_repository'
+require_relative '../validators/ATM_validator'
+require_relative '../validators/account_validator'
+require_relative '../validators/transaction_ATM_validator'
+require_relative '../validators/transaction_validator'
+require_relative '../errors/validation_error'
+require_relative '../model/account'
+require_relative '../model/ATMs'
+require_relative '../model/transaction_ATM'
+require_relative '../model/transaction_account'
+
+class Controller
+  def initialize
+    @transaction_account_repository = TansactionAccountsRepository.instance
+    @transaction_atm_repository = TransactionATMRepository.instance
+    @account_repository = AccountRepository.instance
+    @atm_repository = ATMsRepository.instance
+  end
+
+  # Adds
+  def add_atm(atm_attributes)
+    atm = ATMs.new(atm_attributes[:id], atm_attributes[:location])
+
+    begin
+      ATMValidator.validate(atm)
+      @atm_repository.add_atm(atm)
+    rescue ValidationError => e
+      puts "Error adding ATM: #{e.message}"
+    end
+  end
+
+  def add_account(account_attributes)
+    account = Account.new(account_attributes[:id], account_attributes[:balance], account_attributes[:name],
+                          account_attributes[:job], account_attributes[:email], account_attributes[:address])
+    begin
+      AccountValidator.validate(account)
+      @account_repository.add_account(account)
+    rescue ValidationError => e
+      puts "Error adding account: #{e.message}"
+    end
+  end
+
+  def add_transaction_atm(transaction_atm_attributes)
+    transaction_atm = TransactionATM.new(transaction_atm_attributes[:id], transaction_atm_attributes[:account_id],
+                                         transaction_atm_attributes[:atm_id], transaction_atm_attributes[:amount],
+                                         transaction_atm_attributes[:timestamp], transaction_atm_attributes[:is_withdrawal] || false)
```

Comment:
you already have a default false value in TransactionATM, so you don't need `|| false`

Also, you could probably use the double splat operator
```ruby
TransactionATM.new(**transaction_atm_attributes)
```